tensorflow/core/ops/math_ops.cc

```diff
@@ -2365,4 +2365,132 @@ output_max: the computed max output.
+    .SetShapeFn(ParticalReductionShapeFn)
+    .Doc(R"doc(
+Dynamically sum over the first dimension of a tensor according to start and end
+indices specified at "index". For example if input is
+[[1,1,1],[10,10,10],[100,100,100],[1000,1000,1000]] and index is [[0,1],[1,1],[0,2]],
+the the output will be [[1,1,1],[0,0,0],[11,11,11]]. The data must be at least rank
+one. The indices must be of shape (?,2) where the first column is start indices and
+the second column is end indices. The end index is not included in the sum, which
+means, if you want to do a sum over 0,1,2, the you should have start index 0 and end
+index 3. If end index is smaller than or equal to start, the result will be zero.
+If end index is out of bounds, then the sum will automatically stop at the bound,
+so feel free to put a large number as your end of your index if you want to sum until
+the bound.
+
+data: The source of data where the computation will be taken from.
+indices: start, end indices that controls which part to be included.
+T: the type of data.
+Tindices: the type of indices, must be int32 or int64.
+output: the computed sum values.
+)doc");
+
+REGISTER_OP("PartialProd")
+    .Input("data: T")
+    .Input("indices: Tindices")
+    .Output("output: T")
+    .Attr("T: numbertype")
+    .Attr("Tindices: {int32,int64}")
+    .SetShapeFn(ParticalReductionShapeFn)
+    .Doc(R"doc(
+Dynamically compute the product over the first dimension of a tensor according
+to start and end indices specified at "index". For example if input is
+[[1,2,3],[10,20,30],[100,200,300],[10,10,10]] and index is [[0,1],[1,1],[0,2]],
+the the output will be [[1,2,3],[1,1,1],[10,40,90]]. The data must be at least rank
+one. The indices must be of shape (?,2) where the first column is start indices and
+the second column is end indices. The end index is not included in the product, which
+means, if you want to do a product over 0,1,2, the you should have start index 0 and end
+index 3. If end index is smaller than or equal to start, the result will be one.
+If end index is out of bounds, then the sum will automatically stop at the bound,
+so feel free to put a large number as your end of your index if you want to multiply until
+the bound.
+
+data: The source of data where the computation will be taken from.
+indices: start, end indices that controls which part to be included.
+T: the type of data.
+Tindices: the type of indices, must be int32 or int64.
+output: the computed product values.
+)doc");
+
+REGISTER_OP("PartialMax")
+    .Input("data: T")
+    .Input("indices: Tindices")
+    .Output("output: T")
+    .Attr("T: numbertype")
+    .Attr("Tindices: {int32,int64}")
+    .SetShapeFn(ParticalReductionShapeFn)
+    .Doc(R"doc(
+Dynamically compute the maximum over the first dimension of a tensor according
+to start and end indices specified at "index". For example if input is
+[[1,2,3],[10,20,30],[100,200,300],[10,10,10]] and index is [[0,1],[1,1],[0,2]],
```

Comment:
It might be worth using pretty print to make the matrices multi line so they are more readable see array_ops.cc:884 "MatrixBandPart"